neps/nep-0413.md

```diff
@@ -0,0 +1,153 @@
+An attacker could make the user inadvertently sign a valid transaction which, once signed, could be submitted into the network to execute it.
+
+### Why The Message Needs to Include a Service Identifier?
+To avoid replay attacks, i.e. a malicious app requesting the user to authenticate in their service, only to use the authentication message in a third-app. 
+
+Including the service's identifier (e.g. domain name) and making sure the user knows about it should mitigate this kind of attacks.
+
+### Why a FullAccess Key? Why Not Simply Creating an [FunctionCall Key](https://docs.near.org/concepts/basics/accounts/access-keys) for Signing?
+The most common flow for [NEAR user authentication into a Web3 frontend](https://docs.near.org/develop/integrate/frontend#user-sign-in--sign-out) involves the creation of a [FunctionCall Key](](https://docs.near.org/concepts/basics/accounts/access-keys)).
+
+One might feel tempted to reproduce such process here, for example, by creating a key that can only be used to call a non-existing method in the user's account. This is a bad idea for several reasons:
+1. It implies that the user needs to expend gas in creating a new key.
+2. Since any third-party can ask the user to create a `FunctionCall Key`, this opens a vector of attack.
+
+Using a FullAccess key allows us to be sure that the message was signed by the user (since nobody should have access to their `FullAccess Key`), while keeping the constraints of not expending gas in the process (because no new key needs to be created).
+
+### How to Return the Signed Message in a Safe Way
+Sending the signature in a query string to an arbitrary URL (even within the correct domain) is not secure as the data can be leaked (e.g. through headers, etc). Using URL fragments instead will improve security, since the fragment isn't included in the Referrer. Further constraining the callback URL to a fixed value will improve security even more.
+
+### NEAR Signatures
+NEAR transaction signatures are not plain Ed25519 signatures but Ed25519 signatures of a SHA-256 hash (see [near/nearcore#2835](https://github.com/near/nearcore/issues/2835)). Any protocol that signs anything with NEAR account keys should use the same signature format.
+
+## Specification
+Given the previous [rationales](#rationale-and-alternatives), this NEP proposes the wallets to implement a `verifyOwner` method following the specifications bellow.
+
+
+### Goal
+The `verifyOwner` method must take a plain text `message` and transform it into a signature.
+
+### Input Interface
+In order to sign a message, `verifyOwner` must implement the following input interface:
+
+```jsx
+interface VerifyOwnerParams {
+  message: string; // The message requested to be signed, should contain the app name.
+  callbackUrl?: string; // Applicable to browser wallets (e.g. MyNearWallet). This is the callback url once the signing is approved. Defaults to `window.location.href`.
```

Comment:
One final thing that got my attention. This `callbackUrl` is to be used to get the result. What and how would the parameters look in failed (example: denied) verifying of ownership?

For this case the callback URLs for the sign transaction method have success and failure callback URLs. Maybe we can do the same here.

Or, we can add parameter of `errorCode` as parameter for failure to verify owner.

Reply:
I see that the current implementation returns `<callbackUrl>?error=<error-message-string>`, maybe we could keep that? what do you think @toteto

Reply:
That would work and if that is already how errors are handled by the other methods, I am OK with it since it will be more familiar for the developers.

However, think preferable solution instead of error-mesaage is to include error code since it is bit easier for erroe handling (no need to do weird string comparisons) and also easier for localization. Something like `<callbackUrl>?error=<error-code-string>`, example `<callbackUrl>?error=user-denied`